sdk/program/src/stable_layout/stable_vec.rs

```diff
@@ -0,0 +1,186 @@
+//! `Vec`, with a stable memory layout
+
+use std::{marker::PhantomData, mem::ManuallyDrop, ptr::NonNull};
+
+/// `Vec`, with a stable memory layout
+///
+/// This container is used within the runtime to ensure memory mapping and memory accesses are
+/// valid.  We rely on known addresses and offsets within the runtime, and since `Vec`'s layout
+/// is allowed to change, we must provide a way to lock down the memory layout.  `StableVec`
+/// reimplements the bare minimum of `Vec`'s API sufficient only for the runtime's needs.
+///
+/// To ensure memory allocation and deallocation is handled correctly, it is only possible to
+/// create a new `StableVec` from an existing `Vec`.  This way we ensure all Rust invariants are
+/// upheld.
+///
+/// # Examples
+///
+/// Creating a `StableVec` from a `Vec`
+///
+/// ```
+/// # use solana_program::stable_layout::stable_vec::StableVec;
+/// let vec = vec!["meow", "woof", "moo"];
+/// let vec = StableVec::from(vec);
+/// ```
+#[repr(C)]
+pub struct StableVec<T> {
+    pub ptr: NonNull<T>,
+    pub cap: usize,
+    pub len: usize,
+    _marker: PhantomData<T>,
+}
+
+impl<T> StableVec<T> {
+    #[inline]
+    pub fn as_ptr(&self) -> *const T {
+        // We shadow the slice method of the same name to avoid going through
+        // `deref`, which creates an intermediate reference.
+        self.ptr.as_ptr()
+    }
+
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        // We shadow the slice method of the same name to avoid going through
+        // `deref_mut`, which creates an intermediate reference.
+        self.ptr.as_ptr()
+    }
+}
+
+impl<T> AsRef<[T]> for StableVec<T> {
+    fn as_ref(&self) -> &[T] {
+        self
+    }
+}
+
+impl<T> AsMut<[T]> for StableVec<T> {
+    fn as_mut(&mut self) -> &mut [T] {
+        self
+    }
+}
+
+impl<T> std::ops::Deref for StableVec<T> {
```

Comment:
I don't think we're using these extra trait impls, and if so I'd say let's
remove them

Reply:
Ah uh no this seems worse, I say leave it as is